x/observer/keeper/observer_mapper.go

```diff
@@ -85,9 +86,36 @@ func (k Keeper) GetAllObserverMappersForAddress(ctx sdk.Context, address string)
 
 // Tx
 
-// Not implemented.
 func (k msgServer) AddObserver(goCtx context.Context, msg *types.MsgAddObserver) (*types.MsgAddObserverResponse, error) {
-	_ = sdk.UnwrapSDKContext(goCtx)
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if msg.Creator != k.GetParams(ctx).GetAdminPolicyAccount(types.Policy_Type_add_observer) {
+		return &types.MsgAddObserverResponse{}, types.ErrNotAuthorizedPolicy
+	}
+	observerMappers := k.GetAllObserverMappers(ctx)
+	totalObserverCountCurrentBlock := uint64(0)
+	for _, mapper := range observerMappers {
+		mapper.ObserverList = append(mapper.ObserverList, msg.ObserverAddress)
+		totalObserverCountCurrentBlock += uint64(len(mapper.ObserverList))
+		k.SetObserverMapper(ctx, mapper)
+	}
+	if totalObserverCountCurrentBlock < 0 {
+		return &types.MsgAddObserverResponse{}, types.ErrObserverCountNegative
+	}
+	pubkey, _ := common.NewPubKey(msg.ZetaclientGranteePubkey)
```

Comment:
I think we should still handle the error even if it is handled in message validation

Reply:
Can you think of a way in which validate basic can be skipped by a TX? I know it s a minute operation , but if not needed we can skip it , what are your thoughts ?

Reply:
We have some message handlers in the code that can be called without sending a message, wondering if it could be the case for this one. But actually even outside of this, we are supposed to consider the validity of the message a precondition for the message handler execution, so it's fine from my side.

Reply:
i agree with @lumtis , it takes **much less time** to add the error check here than all three of us trying to **think** about it. Maybe @kingpinXD you are right, but to prevent future waste of time, attention, and energy, every error in zetacored needs to be checked.